modules/db/src/main/scala/gem/dao/ObservationDao.scala

```diff
@@ -70,11 +70,12 @@ object ObservationDao {
   /** Construct a program to select a fully specified observation, with targets,
     * static config and steps.
     */
+  @SuppressWarnings(Array("org.wartremover.warts.AsInstanceOf"))
   def select(id: Observation.Id): ConnectionIO[Observation.Full] =
     for {
       o <- selectConfig(id)
       t <- TargetEnvironmentDao.selectObs(id, o.targets)
-    } yield Observation.targetsFunctor.as(o, t)
+    } yield Observation.targetsFunctor.as(o, t).asInstanceOf[Observation.Full] // TODO: push this assertion lower
```

Comment:
Seems like we'll usually be working with an `Observation.Full` with unknown instrument.  I guess I don't *know* that will be the case but how could we know a priori what the instrument will be.

Reply:
In the comprehension above `o` knows the instrument so we could push this down into `selectObs` as an expectation that we check, and if it lines up we can perform the cast down there, such that `o` and `t` line up. Maybe.

Anyway in the same way we assert the types of database rows and blow up (programmer error!) if we're wrong, we need to check that the parts line up and then perform the cast so *Scala* will know. And we want to do this as quickly as possible so the largest amount of code knows it.